doc/reference/trd103-gpio.md

```diff
@@ -0,0 +1,175 @@
+
+Abstract
+-------------------------------
+
+This document describes the hardware independent layer interface (HIL) for
+General Purpose Input/Output (GPIO) in the Tock operating system kernel.  It
+describes the Rust traits and other definitions for this service as well as the
+reasoning behind them. This document is in full compliance with [TRD1].
+
+1. Introduction
+-------------------------------
+
+General Purpose Input/Output (GPIO) controls generic pins. User code can control
+the output level on the pin (high or low), read the externally drive logic level
+and often configure pull-up or pull-down resistence. Typically, microcontrollers
+expose pins in groups called ports however Tock's GPIO HIL exposes pins
+individually since ports often do not group pins as they are actually used on a
+board. Software that wishes to control a whole port (e.g. for efficiency) should
+use the per-chip implementation, which may export this feature.
+
+The GPIO HIL is the kernel crate, in module hil::gpio. It provides three traits:
+
+  * `kernel::hil::gpio::Pin`: Controls and reads output level and
+    enables/disables interrupts on a single pin.
+  * `kernel::hil::gpio::PinCtl`: Controlls the input mode on a single pin.
+  * `kernel::hil::gpio::Client`: handles the callback when a GPIO interrupt is
+    fired.
+
+The rest of this document discusses each in turn.
+
+2. `Pin` trait
+-------------------------------
+
+The AdcSingle trait is for requesting a single ADC conversion. It has
+three functions:
+
+```rust
+pub trait Pin {
+    /// Configure the GPIO pin as an output pin.
+    fn make_output(&self);
+
+    /// Configure the GPIO pin as an input pin.
+    fn make_input(&self);
+
+    /// Disable the GPIO pin and put it into its lowest power
+    /// mode.
+    fn disable(&self);
+
+    /// Set the GPIO pin high. It must be an output.
+    fn set(&self);
+
+    /// Set the GPIO pin low. It must be an output.
+    fn clear(&self);
+
+    /// Toggle the GPIO pin. It must be an output.
+    fn toggle(&self);
+
+    /// Get the current state of an input GPIO pin.
+    fn read(&self) -> bool;
+
+    /// Enable an interrupt on the GPIO pin. It must
+    /// be configured as an interrupt. The `identifier`
+    /// can be any value and will be returned to you
+    /// when the interrupt on this pin fires.
+    fn enable_interrupt(&self, identifier: usize, mode: InterruptMode);
+
+    /// Disable the interrupt for the GPIO pin.
+    fn disable_interrupt(&self);
+}
+```
+
+Either the `make_output` or `make_input` methods MUST be called at least once
+before any other methods are called. The `make_output` method MUST put the pin
+into output mode and ensure that `set`, `clear` and `toggle` are effective. The
+`make_input` method MUST put the pin into input mode and ensure that `read`
+correctly returns the logic level of the pin. It is OPTIONAL whether input
+methods are supported in output mode and whether output methods are supported in
+input mode.
```

Comment:
The OPTIONAL means there needs to be an is_input and an is_output, otherwise you can't tell whether make_input has undone a call to make_output. There should probably be a function to check whether simultaneous input/output is allowed?